Content.Shared/Burial/BurialSystem.cs

```diff
@@ -0,0 +1,176 @@
+
+        if (TryComp<ToolComponent>(args.Used, out var shovel) && _toolSystem.HasQuality(args.Used, "Digging"))
+        {
+            _doAfterSystem.TryStartDoAfter(new DoAfterArgs(EntityManager, args.User, component.DigDelay / shovel.SpeedModifier, new GraveDiggingDoAfterEvent(), uid, target: args.Target, used: uid)
+            {
+                BreakOnTargetMove = true,
+                BreakOnUserMove = true,
+                BreakOnDamage = true,
+                NeedHand = true,
+                BreakOnHandChange = true
+            });
+
+            StartDigging(uid, args.User, args.Used, component);
+        }
+        else
+        {
+            _popupSystem.PopupClient(Loc.GetString("grave-digging-requires-tool", ("grave", args.Target)), uid, args.User);
+        }
+    }
+
+    private void OnAfterInteractUsing(EntityUid uid, GraveComponent component, AfterInteractUsingEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        // don't place shovels on the grave, only dig
+        if (_toolSystem.HasQuality(args.Used, "Digging"))
+            args.Handled = true;
+    }
+
+    private void OnActivate(EntityUid uid, GraveComponent component, ActivateInWorldEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        if(_gameTiming.IsFirstTimePredicted)
+            _popupSystem.PopupClient(Loc.GetString("grave-digging-requires-tool", ("grave", args.Target)), uid, args.User);
+    }
+
+    private void OnGraveDigging(EntityUid uid, GraveComponent component, GraveDiggingDoAfterEvent args)
+    {
+        RemComp<ActiveGraveComponent>(uid);
+
+        if (args.Cancelled || args.Handled)
+            return;
+
+        component.DiggingComplete = true;
+
+        if (args.Used != null)
+            _storageSystem.ToggleOpen(args.User, uid);
+        else
+            _storageSystem.TryOpenStorage(args.User, uid); //can only claw out
+    }
+
+    private void StartDigging(EntityUid uid, EntityUid user, EntityUid? used, GraveComponent component)
+    {
+        ActiveGraveComponent activeComp = EnsureComp<ActiveGraveComponent>(uid);
+
+        if (_gameTiming.IsFirstTimePredicted)
+        {
+            if (used != null)
+            {
+                _popupSystem.PopupEntity(Loc.GetString("grave-start-digging-others", ("user", user), ("grave", uid), ("tool", used)), user, Filter.PvsExcept(user), true);
+                if (_gameTiming.InPrediction)
+                {
+                    _popupSystem.PopupEntity(Loc.GetString("grave-start-digging-user", ("grave", uid), ("tool", used)), user, user);
+                    activeComp.Stream = _audioSystem.PlayPvs(component.DigSound, uid).Value.Entity;
+                }
+            }
+            else
+            {
+                activeComp.DiggingSelfOut = true;
+                _popupSystem.PopupClient(Loc.GetString("grave-start-digging-user-trapped", ("grave", uid)), user, user, PopupType.Medium);
+            }
+        }
+    }
```

Comment:
This method needs cleaning up, just use the predicted methods internally and remove the outer wrapper of IsFirstTimePredicted for it.

Reply:
After making it a NetworkedComponent, there were some issues using PlayPredicted with a loop. It would never seem to stop the loop properly. I tried many different things, but what seemed to work is using PlayPvs after checking IsServer, like we currently do in SharedSolutionContainerMixerSystem. If there's a better way to do this, just let me know.